engine/src/internal_development.cpp

```diff
@@ -494,61 +494,78 @@ struct MCObjectListener
-	// MW-2013-08-27: [[ Bug 11126 ]] If this pointer is going away, make sure we fetch the next
-	//   listener into the static.
-	if (s_next_listener_to_process == p_listener)
-		s_next_listener_to_process = p_listener -> next;
-	MCMemoryDelete(p_listener);
-	p_listener = p_prev_listener;
-}
+    if (s_listeners_locked)
+        return;
+    
+    MCObjectListener *t_prev_listener;
+    t_prev_listener = nil;
+    
+    MCObjectListener *t_listener;
+    t_listener = s_object_listeners;
 
-static void remove_object_listener_target_from_list(MCObjectListenerTarget *&p_target, MCObjectListenerTarget *p_prev_target, MCObjectListener* &p_listener, MCObjectListener* p_prev_listener)
-{
-	p_target -> target -> Release();
-	if (p_prev_target != nil)
-		p_prev_target -> next = p_target -> next;
-	else
-		p_listener -> targets = p_target -> next;
-	if (p_listener -> targets == nil)
-		remove_object_listener_from_list(p_listener, p_prev_listener);
-	// MW-2013-08-27: [[ Bug 11126 ]] If this pointer is going away, make sure we fetch the next
-	//   target into the static.
-	if (p_target == s_next_listener_target_to_process)
-		s_next_listener_target_to_process = p_target -> next;
-	MCMemoryDelete(p_target);
-	p_target = p_prev_target;
+    for (t_listener = s_object_listeners; t_listener != nil; t_listener = t_listener -> next)
+    {
+        MCObjectListenerTarget *t_target;
+        t_target = nil;
+        MCObjectListenerTarget *t_prev_target;
+        t_prev_target = nil;
+        
+        bool t_listener_exists;
+        t_listener_exists = t_listener -> object -> Exists();
+        
+        // Check all the listener's targets to see if they can be pruned
+        for (t_target = t_listener -> targets; t_target != nil; t_target = t_target -> next)
+        {
+            // Prune target from listener target list if listener
+            // doesn't exist, or if target doesn't exist
+            if (!t_listener_exists || !t_target -> target -> Exists())
+            {
+                t_target -> target -> Release();
+                
+                if (t_prev_target != nil)
+                    t_prev_target -> next = t_target -> next;
+                else
+                    t_listener -> targets = t_target -> next;
+                
+                MCMemoryDelete(t_target);
+                t_target = t_prev_target;
+            }
+        }
+        
+        // If the listening object no longer exists, or the list of
+        // listener targets is nil, then prune from listener list.
+        if (!t_listener_exists || t_listener -> targets == nil)
+        {
+            t_listener -> object -> Release();
+            
+            if (t_prev_listener != nil)
+                t_prev_listener -> next = t_listener -> next;
+            else
+                s_object_listeners = t_listener -> next;
+            
+            MCMemoryDelete(t_listener);
+            t_listener = t_prev_listener;
+        }
```

Comment:
I think t_prev_listener needs updated here - if you aren't deleting the current one.